[BugFix][torch.compile] Fix fused_scaled_matmul_reduce_scatter signature for PyTorch 2.8

jasonlizhengjian opened the pull request:
## Purpose
Fixes #24376 part 2.d

Updated torch.ops.symm_mem.fused_scaled_matmul_reduce_scatter calls to match the new PyTorch API signature. The function signature changed from PyTorch 2.7.1 to require additional positional parameters.

Changes:
- Added orig_scatter_dim and scatter_dim_after_maybe_reshape as positional parameters
- Added output_shape calculation: [*input.shape[:-1], mat2.shape[1]]
- Changed all optional parameters (bias, result_scale, out_dtype, use_fast_accum) from keyword arguments to positional arguments to match PyTorch's torch._inductor implementation

References:
- PyTorch function definition: [torch/distributed/_symmetric_memory/__init__.py](https://github.com/pytorch/pytorch/blob/main/torch/distributed/_symmetric_memory/__init__.py#L454)
- PyTorch test usage: [test/distributed/test_symmetric_memory.py](https://github.com/pytorch/pytorch/blob/main/test/distributed/test_symmetric_memory.py#L578)
- PyTorch inductor usage: [torch/_inductor/fx_passes/micro_pipeline_tp.py](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/fx_passes/micro_pipeline_tp.py#L818)

## Test Plan

`tests/compile/test_async_tp.py`

## Test Result

`pytest -s tests/compile/test_async_tp.py` on 2 x B200
```
================== 16 passed, 9 warnings in 350.92s (0:05:50) ==================
```

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

jasonlizhengjian committed:
Fix fused_scaled_matmul_reduce_scatter signature for PyTorch update

Updated torch.ops.symm_mem.fused_scaled_matmul_reduce_scatter calls to match
the new PyTorch API signature. The function signature changed from PyTorch 2.7.1
to require additional positional parameters.

Changes:
- Added orig_scatter_dim and scatter_dim_after_maybe_reshape as positional parameters
- Added output_shape calculation: [*input.shape[:-1], mat2.shape[1]]
- Changed all optional parameters (bias, result_scale, out_dtype, use_fast_accum)
  from keyword arguments to positional arguments to match PyTorch's torch._inductor
  implementation

References:
- PyTorch function definition: torch/distributed/_symmetric_memory/__init__.py:454-461
- PyTorch test usage: test/distributed/test_symmetric_memory.py:579-590
- PyTorch inductor usage: torch/_inductor/fx_passes/micro_pipeline_tp.py:816-834

Signed-off-by: jasonlizhengjian <jasonlizhengjian@gmail.com>

jasonlizhengjian commented:
@cascade812 can you also review

ProExpertProg reviewed (approved):
Can we add the test file to PyTorch Compilation Unit Tests in `test-pipeline.yml`?

jasonlizhengjian commented:
> Can we add the test file to PyTorch Compilation Unit Tests in `test-pipeline.yml`?

@ProExpertProg 

`test_async_tp.py` is already there but it's currently being skipped due to only being run with 1 GPU:
```
- label: PyTorch Compilation Unit Tests # 15min
  timeout_in_minutes: 30
  mirror_hardwares: [amdexperimental]
  torch_nightly: true
  source_file_dependencies:
    - vllm/
    - tests/compile
  commands:
    - pytest -v -s compile/test_pass_manager.py
    - pytest -v -s compile/test_fusion.py
    - pytest -v -s compile/test_fusion_attn.py
    - pytest -v -s compile/test_silu_mul_quant_fusion.py
    - pytest -v -s compile/test_sequence_parallelism.py
    - pytest -v -s compile/test_async_tp.py
    - pytest -v -s compile/test_fusion_all_reduce.py
    - pytest -v -s compile/test_decorator.py
    - pytest -v -s compile/test_noop_elimination.py
```
for example in this ci run https://buildkite.com/vllm/ci/builds/32809/steps/canvas?jid=019995c2-21ec-4c60-ba64-5052e2ada32f it outputs:

```
[2025-09-29T14:42:06Z] compile/test_async_tp.py::test_async_tp_pass_replace[dtype0-16-16-8-TestScaledMMRSModel] SKIPPED
[2025-09-29T14:42:06Z] compile/test_async_tp.py::test_async_tp_pass_replace[dtype0-16-16-8-TestAGScaledMMModel] SKIPPED
[2025-09-29T14:42:06Z] compile/test_async_tp.py::test_async_tp_pass_replace[dtype0-16-16-8-TestCutlassScaledMMRSModel] SKIPPED
[2025-09-29T14:42:06Z] compile/test_async_tp.py::test_async_tp_pass_replace[dtype0-16-16-8-TestAGCutlassScaledMMModel] SKIPPED
[2025-09-29T14:42:06Z] compile/test_async_tp.py::test_async_tp_pass_replace[dtype1-16-16-8-TestMMRSModel] SKIPPED
[2025-09-29T14:42:06Z] compile/test_async_tp.py::test_async_tp_pass_replace[dtype1-16-16-8-TestAGMMModel] SKIPPED
[2025-09-29T14:42:06Z] compile/test_async_tp.py::test_async_tp_pass_replace[dtype1-16-16-8-TestScaledMMRSModel] SKIPPED
[2025-09-29T14:42:06Z] compile/test_async_tp.py::test_async_tp_pass_replace[dtype1-16-16-8-TestAGScaledMMModel] SKIPPED
[2025-09-29T14:42:06Z] compile/test_async_tp.py::test_async_tp_pass_replace[dtype1-16-16-8-TestCutlassScaledMMRSModel] SKIPPED
[2025-09-29T14:42:06Z] compile/test_async_tp.py::test_async_tp_pass_replace[dtype1-16-16-8-TestAGCutlassScaledMMModel] SKIPPED
[2025-09-29T14:42:06Z] compile/test_async_tp.py::test_async_tp_pass_correctness[False-mp-True-2-meta-llama/Llama-3.2-1B-Instruct] Fork a new process to run a test 772
[2025-09-29T14:42:06Z] Fork a new process to run a test 0
[2025-09-29T14:42:06Z] Need at least 2 x 1 GPUs
[2025-09-29T14:42:07Z] PASSED
[2025-09-29T14:42:07Z] compile/test_async_tp.py::test_async_tp_pass_correctness[False-mp-True-2-RedHatAI/Meta-Llama-3.1-8B-Instruct-FP8] Fork a new process to run a test 773
[2025-09-29T14:42:07Z] Fork a new process to run a test 0
[2025-09-29T14:42:07Z] Need at least 2 x 1 GPUs
[2025-09-29T14:42:07Z] PASSED
[2025-09-29T14:42:07Z] compile/test_async_tp.py::test_async_tp_pass_correctness[True-mp-True-2-meta-llama/Llama-3.2-1B-Instruct] Fork a new process to run a test 774
[2025-09-29T14:42:07Z] Fork a new process to run a test 0
[2025-09-29T14:42:07Z] Need at least 2 x 1 GPUs
[2025-09-29T14:42:07Z] PASSED
[2025-09-29T14:42:07Z] compile/test_async_tp.py::test_async_tp_pass_correctness[True-mp-True-2-RedHatAI/Meta-Llama-3.1-8B-Instruct-FP8] Fork a new process to run a test 775
[2025-09-29T14:42:07Z] Fork a new process to run a test 0
[2025-09-29T14:42:07Z] Need at least 2 x 1 GPUs
[2025-09-29T14:42:07Z] PASSED
```
Not sure why some of these say PASSED despite having the output from the skip:
```
    if num_gpus_available < tp_size:
        pytest.skip(f"Need at least {tp_size} x {pp_size} GPUs")
 ```

andoorve commented:
cc: @louiswang524 re: https://github.com/vllm-project/vllm/pull/24393

cascade812 commented:
@jasonlizhengjian thanks, LGTM

tejas-srikanth commented:
@jasonlizhengjian Thank you for taking care of this, LGTM!

jasonlizhengjian committed:
Move test_async_tp.py to Distributed Tests (2 GPUs) section

Moved compile/test_async_tp.py from Compilation Tests to Distributed Tests
(2 GPUs) section as it requires 2 GPUs to run (@multi_gpu_test decorator).

Also added tests/compile/test_async_tp.py to source_file_dependencies.

Signed-off-by: jasonlizhengjian <jasonlizhengjian@gmail.com>

jasonlizhengjian committed:
Apply ruff-format to collective_fusion.py after rebase

Signed-off-by: Jason Li <jasonlizhengjian@gmail.com>

Signed-off-by: jasonlizhengjian <jasonlizhengjian@gmail.com>

jasonlizhengjian committed:
Move test_sequence_parallelism.py to Distributed Tests (2 GPUs) section

Moved compile/test_sequence_parallelism.py from Compilation Tests to
Distributed Tests (2 GPUs) section as it requires 2 GPUs to run
(@multi_gpu_test decorator).

Also added tests/compile/test_sequence_parallelism.py to
source_file_dependencies.

Signed-off-by: jasonlizhengjian <jasonlizhengjian@gmail.com>

jasonlizhengjian committed:
Move test_async_tp.py to H200 test section

The test_async_tp.py uses PyTorch's symmetric memory operations
(torch.ops.symm_mem.fused_matmul_reduce_scatter and
fused_all_gather_matmul) which require Hopper architecture (H100/H200)
or newer GPUs with multicast and fabric support.

L4 GPUs (Ada Lovelace, compute capability 8.9) lack these features,
causing "invalid device ordinal" errors during symmetric memory
rendezvous when the test runs on the default L4 runners.

Changes:
- Removed test_async_tp.py from "Distributed Tests (2 GPUs)" section
  which runs on L4 GPUs
- Added test_async_tp.py to "Distributed Tests (H200)" section which
  runs on H200 GPUs with full symmetric memory support
- test_sequence_parallelism.py remains on L4 as it uses standard NCCL
  collectives that don't require symmetric memory

Signed-off-by: jasonlizhengjian <jasonlizhengjian@gmail.com>

Signed-off-by:  <>

jasonlizhengjian commented:
@ProExpertProg the Async TP test was failing in distributed tests https://buildkite.com/vllm/ci/builds/33859/steps/canvas?sid=0199bf91-e585-47ed-adff-0300604a3c12#0199bf91-e674-4667-b280-e1c472cebec7/165-4049 
since it seemed like there was some issue with torch symmetric memory and L4 GPUs. 
I moved it to the H200 section for now

jasonlizhengjian committed:
Move test_async_tp.py and test_sequence_parallelism.py to H200 test section

Both test_async_tp.py and test_sequence_parallelism.py use PyTorch's
symmetric memory operations which require Hopper architecture (H100/H200)
or newer GPUs with multicast and fabric support.

L4 GPUs (Ada Lovelace, compute capability 8.9) lack these features,
causing "invalid device ordinal" errors during symmetric memory
rendezvous when the tests run on the default L4 runners.

Changes:
- Removed test_async_tp.py from "Distributed Tests (2 GPUs)" section
  which runs on L4 GPUs
- Removed test_sequence_parallelism.py from "Distributed Tests (2 GPUs)"
  section which runs on L4 GPUs
- Added both tests to "Distributed Tests (H200)" section which runs on
  H200 GPUs with full symmetric memory support

Signed-off-by: jasonlizhengjian <jasonlizhengjian@gmail.com>

Signed-off-by:  <>

jasonlizhengjian commented:
@ProExpertProg I had to move seq parallel test to the H200 section as well due to the fp8 parts

ProExpertProg committed:
Reorder tests

Signed-off-by: Luka Govedič <ProExpertProg@users.noreply.github.com>
